execution_engine/src/storage/global_state/mod.rs

```diff
@@ -123,6 +126,401 @@ pub trait StateProvider {
+    where
+        S: TrieStore<K, V>,
+        T: Readable<Handle = S::Handle>,
+        K: FromBytes,
+        V: FromBytes,
+        S::Error: From<T::Error>,
+    {
+        Ok(
+            match store
+                .get(tx, digest)?
+                // NOTE: Should we map to an error here?
+                .expect("Hashes referred to must be in global state.")
+            {
+                Trie::Leaf { key, value } => Pointer::new_leaf(key, value),
+                Trie::Extension { affix, pointer } => Pointer::new_extension(
+                    &affix,
+                    Pointer {
+                        link: Link::GlobalState(*pointer.hash()),
+                        is_leaf: matches!(pointer, TriePointer::LeafPointer(_)),
+                    },
+                ),
+                Trie::Node { pointer_block } => {
+                    let mut branches = new_branches();
+                    for (i, p) in pointer_block.as_indexed_pointers() {
+                        branches[i as usize] = Some(Pointer {
+                            link: Link::GlobalState(*p.hash()),
+                            is_leaf: matches!(p, TriePointer::LeafPointer(_)),
+                        });
+                    }
+                    Pointer {
+                        link: Link::Fancy(Box::new(FancyTrie::Node { branches })),
+                        is_leaf: false,
+                    }
+                }
+            },
+        )
+    }
+
+    impl<K, V> Pointer<K, V> {
+        /// Inserts a key-value pair into a fancy trie.
+        /// A new root node may result, which is why this mutates a `Pointer`.
+        //
+        // NOTE: Couldn't make it safe because of how Rust handles lifetime constraints.
+        // We attempted `let mut current: &mut Pointer<K, V> = self`, but then we couldn't assign to
+        // `*current` because `link` and `fancy_box` are borrowed. The assignment to `current` in the
+        // loop imply that the lifetimes of `link` and `fancy_box` have to include that of `current`.
+        // We can't make this recursive either because Rust doesn't guarantee tail call elimination.
+        // This is really regrettable and a safe, non-recursive, version is very welcome.
+        pub fn insert<S, T, E>(&mut self, store: &S, tx: &T, key: K, value: V) -> Result<(), E>
+        where
+            K: FromBytes + ToBytes + PartialEq + std::fmt::Debug,
+            V: FromBytes,
+            S: TrieStore<K, V>,
+            T: Readable<Handle = S::Handle>,
+            S::Error: From<T::Error>,
+            E: From<S::Error> + From<bytesrepr::Error>,
+        {
+            /// The suffix of the key bytes at the node we are considering.
+            let mut bytes: &[u8] = &key.to_bytes()?;
+            /// The node we are considering. It's actually a reference to a pointer to that node
+            /// because we might have to replace it.
+            let mut current: std::ptr::NonNull<Pointer<K, V>> = self.into();
```

Comment:
Unsafe is a sore in my eye and if the problem is lifetimes in recursive data structure I think it's worth giving a `RefCell` a try and turn compile-time borrow check into a runtime one. I.e. `Option<RefCell<Pointer<K, V>>>`?

Reply:
With the obvious downside being overhead, this is a possible solution. However, I think there is little reason to believe this can't be made in safe Rust with the types as they are because it can be done with recursion. I just don't know how.

Reply:
It seems Rust's design so far isn't flexible enough to accommodate a safe implementation of this. Check out this RFC: https://rust-lang.github.io/rfcs/2094-nll.html#problem-case-4-mutating-mut-references